nutils/_util.py

```diff
@@ -778,4 +778,171 @@ def wrapper(*args, **kwargs):
     return wrapper
 
 
+def _reduce(obj):
+    'helper function for deep_replace_property and shallow_replace'
+
+    if obj.__class__ in (tuple, list, dict, set, frozenset):
+        if not obj: # empty containers need not be entered
+            raise Exception('refusing to enter empty container')
+        return lambda *x, T=type(obj): T(x), tuple(obj if not isinstance(obj, dict) else obj.items())
+    return obj.__reduce__()
+
+
+class deep_replace_property:
+    '''decorator for deep object replacement
+
+    Generates a cached property for deep replacement of reduceable objects,
+    based on a callable that is applied depth first and recursively on
+    individual constructor arguments. Intermediate values are stored in the
+    attribute by the same name of any object that is a decendent of the class
+    that owns the property.
+
+    Args
+    ----
+    func
+        Callable which maps an object onto a new object, or `None` if no
+        replacement is made. It must have precisely one positional argument for
+        the object.
+    '''
+
+    def __init__(self, func):
+        self.func = func
+
+    def __set_name__(self, owner, name):
+        self.owner = owner
+        self.name = name
+
+    def __set__(self, obj, value):
+        raise AttributeError("can't set attribute")
+
+    def __delete__(self, obj):
+        raise AttributeError("can't delete attribute")
+
+    def __get__(self, obj, objtype=None):
+        fstack = [obj] # stack of unprocessed objects and command tokens
+        rstack = [] # stack of processed objects
+        recreate = [] # stack of constructor functions
+        originals = [] # stack of original objects to cache new value into
+
+        while fstack:
+            obj = fstack.pop()
+
+            if obj is recreate: # recreate object from rstack
+                f, nargs = recreate.pop()
+                r = f(*[rstack.pop() for _ in range(nargs)])
```

Comment:
Suggestion: Use a generator instead of list comprehension because the byte code of the former is [much shorter](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:python,selection:(endColumn:28,endLineNumber:2,positionColumn:28,positionLineNumber:2,selectionStartColumn:28,selectionStartLineNumber:2,startColumn:28,startLineNumber:2),source:'sum(*(i+for+i+in+range(5)))%0Asum(*%5Bi+for+i+in+range(5)%5D)'),l:'5',n:'0',o:'Python+source+%231',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:python312,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:python,libs:!(),options:'',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+Python+3.12+(Editor+%231)',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4).

Reply:
Second inspection of the byte code reveals that the generator-variant creates a generator function while the list-variant builds a list inline.